File: projects/17-global-conferencing/src/placement.rs
//! V1 — Global room placement via consensus. `src/placement.rs`.
//!
//! This is the control plane every SFU in the mesh must agree on. When a room is first created,
//! *some* region is chosen as its **home** — the anchor where the publisher's origin media lives
//! and where the cascade tree (V2/V3) roots — and that decision must be **identical on every
//! node**, or two people creating the same room from opposite sides of the planet get two
//! disjoint conferences under one id (a *split room*). Deciding one value, cluster-wide, under
//! concurrency and partition, is a **consensus** problem — exactly what your Raft work in
//! **project 09** is for.
//!
//! You're **leaning on** project 09's ideas (randomized election timeout → elect a leader,
//! append-entries → replicate, a commit index, idempotent apply), not rebuilding Raft from
//! scratch. What's *yours* here is modelling **room placement + region membership** as the
//! replicated state machine and getting the safety property right: **at most one home region per
//! room, cluster-wide, forever** — and a minority partition that *cannot* invent a new one.
//!
//! Scaffold state: [`Placement::new`] and the read-only [`Placement::snapshot`] / [`role`] /
//! [`leader`] are wired so the server boots and `/status` shows an (empty) map. The consensus
//! methods — placing a room, replicating membership, the vote/append RPC handlers, and the
//! election tick — are the V1 `todo!()` worklist.

use std::collections::{BTreeSet, HashMap};
use std::sync::RwLock;
use std::time::Duration;

use serde::{Deserialize, Serialize};

use crate::error::Result;

/// A peer SFU in the mesh: its region label, its HTTP control base (for the `/cluster/*` Raft-lite
/// RPCs), and its cascade UDP address (where V2 relays media). Parsed from `PEERS` in `main`.
#[derive(Clone, Debug, Serialize)]
pub struct PeerNode {
    pub region: String,
    /// e.g. `http://10.0.0.2:8080` — where this node POSTs vote/replicate RPCs.
    pub control_addr: String,
    /// e.g. `10.0.0.2:7100` — where this node sends relay media (V2).
    pub media_addr: String,
}

/// This node's role in the Raft-lite placement group. Same three states as project 09.
#[derive(Clone, Copy, Debug, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    Follower,
    Candidate,
    Leader,
}

impl Role {
    /// Numeric encoding for the `conf_node_role` gauge (0 follower, 1 candidate, 2 leader).
    pub fn as_metric(self) -> f64 {
        match self {
            Role::Follower => 0.0,
            Role::Candidate => 1.0,
            Role::Leader => 2.0,
        }
    }
}

/// The committed placement for one room: its single home region and the set of regions with
/// live participants. This is the replicated state machine's value for a room — every node
/// converges to the same `RoomPlacement` for a given `room_id`.
#[derive(Clone, Debug, Serialize)]
pub struct RoomPlacement {
    pub room_id: String,
    /// The one home region, chosen once by consensus. The invariant V1 protects.
    pub home_region: String,
    /// Regions with ≥1 live participant — the cascade topology (V2/V3) is derived from this.
    pub active_regions: BTreeSet<String>,
    /// Bumped on each committed change to this room's placement (membership included).
    pub epoch: u64,
}

/// One entry in the replicated log: the placement state machine's inputs. Applying committed
/// entries in log order yields the same room map on every node (project 09's apply step).
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PlacementEntry {
    /// Claim `region` as the home for a not-yet-placed room. Consensus makes exactly one such
    /// claim win per room; later claims are no-ops (idempotent placement).
    PlaceRoom { room_id: String, region: String },
    /// A region gained its first / lost its last participant in a room — replicated membership.
    RegionInterest {
        room_id: String,
        region: String,
        joined: bool,
    },
}

/// Config for the placement group, read from env in `main`.
pub struct PlacementConfig {
    /// This node's region (the home it proposes for rooms created locally).
    pub region: String,
    /// This node's stable id (used in vote records + logs).
    pub node_id: String,
    /// The other SFUs in the mesh (consensus peers).
    pub peers: Vec<PeerNode>,
    /// Randomized election-timeout base; a follower stands after a random interval in
    /// `[election_timeout, 2*election_timeout)` (project 09's split-vote breaker).
    pub election_timeout: Duration,
    /// Leader heartbeat interval (comfortably below `election_timeout`).
    pub heartbeat: Duration,
    /// Cluster-wide cap on placed rooms, enforced through the log (not per-node).
    pub max_rooms: usize,
}

impl PlacementConfig {
    /// Quorum size for the mesh (self + peers). A minority below this cannot place a room.
    pub fn quorum(&self) -> usize {
        (self.peers.len() + 1) / 2 + 1
    }
}

/// Mutable consensus + applied-state, guarded by one lock.
#[derive(Default)]
struct Inner {
    role: Role,
    /// Current Raft-lite term.
    term: u64,
    /// Who this node believes is leader (region/node id), if any.
    leader: Option<String>,
    /// The applied placement state machine: room id → its committed placement. Rebuilt by
    /// applying the replicated log in order; the source of truth for routing reads.
    rooms: HashMap<String, RoomPlacement>,
}

impl Default for Role {
    fn default() -> Self {
        Role::Follower
    }
}

/// The placement control plane: a Raft-lite replicated map of room → home region + membership.
pub struct Placement {
    cfg: PlacementConfig,
    inner: RwLock<Inner>,
}

impl Placement {
    /// Build the placement group. Wiring only — no election runs until [`run`](Self::run).
    pub fn new(cfg: PlacementConfig) -> Self {
        Self {
            cfg,
            inner: RwLock::new(Inner::default()),
        }
    }

    pub fn config(&self) -> &PlacementConfig {
        &self.cfg
    }

    /// This node's current role (for `/status` + the role gauge).
    pub fn role(&self) -> Role {
        self.inner.read().expect("placement lock").role
    }

    /// This node's current term.
    pub fn term(&self) -> u64 {
        self.inner.read().expect("placement lock").term
    }

    /// Who this node believes leads the placement group, if anyone.
    pub fn leader(&self) -> Option<String> {
        self.inner.read().expect("placement lock").leader.clone()
    }

    /// A read-only view of every placed room (for `/status` + `/rooms`). Comes from the
    /// locally-applied map — no cross-region round-trip (the "placement map is the read cache").
    pub fn snapshot(&self) -> Vec<RoomPlacement> {
        self.inner
            .read()
            .expect("placement lock")
            .rooms
            .values()
            .cloned()
            .collect()
    }

    /// The committed placement for one room, if placed (a hot routing read).
    pub fn room(&self, room_id: &str) -> Option<RoomPlacement> {
        self.inner
            .read()
            .expect("placement lock")
            .rooms
            .get(room_id)
            .cloned()
    }

    // ---- V1 worklist: consensus (elect · replicate · apply) ------------------------------

    /// TODO(V1): Ensure `room_id` has a home region, cluster-wide. If it's unplaced, propose a
    /// `PlaceRoom` entry (home = this node's region) and drive it through consensus; if this node
    /// isn't leader, forward to the leader (or return [`AppError::NotLeader`]). **Idempotent**:
    /// an already-placed room returns its existing placement with no second home and no epoch
    /// churn. A minority partition (below [`PlacementConfig::quorum`]) must **refuse**
    /// ([`AppError::Unavailable`]) rather than invent a home — that refusal is what prevents a
    /// split room. Enforces `max_rooms` through the log.
    pub async fn place_room(&self, room_id: &str) -> Result<RoomPlacement> {
        let _ = (
            room_id,
            &self.cfg.region,
            self.cfg.max_rooms,
            self.cfg.quorum(),
        );
        todo!("V1: place the room via consensus (idempotent, one home, minority refuses)")
    }

    /// TODO(V1): Replicate that `region` gained its first (`joined`) or lost its last
    /// (`!joined`) participant in `room_id`, so every node's `active_regions` for that room
    /// converges. Committed through the same log as placement; drives V2/V3 topology.
    pub async fn register_interest(&self, room_id: &str, region: &str, joined: bool) -> Result<()> {
        let _ = (room_id, region, joined);
        todo!("V1: replicate the membership change, apply on commit")
    }

    /// TODO(V1): Handle an inbound RequestVote-style RPC from a peer (`POST /cluster/vote`).
    /// Grant/deny per the term + log-freshness rules from project 09; update term/role.
    pub async fn on_vote(&self, from: &str, term: u64) -> Result<bool> {
        let _ = (from, term);
        todo!("V1: RequestVote — grant at most one vote per term, adopt higher terms")
    }

    /// TODO(V1): Handle an inbound AppendEntries-style RPC from the leader
    /// (`POST /cluster/replicate`): a heartbeat (empty) or new [`PlacementEntry`]s to append +
    /// commit. Reset the election timer, apply newly-committed entries to the room map.
    pub async fn on_append(
        &self,
        from: &str,
        term: u64,
        entries: Vec<PlacementEntry>,
    ) -> Result<bool> {
        let _ = (from, term, entries);
        todo!("V1: AppendEntries — append/commit entries, apply to the map, reset election timer")
    }

    /// TODO(V1): The election + heartbeat loop. As a follower, stand for election on a
    /// randomized timeout (split-vote breaker from p09); as leader, send heartbeats every
    /// [`PlacementConfig::heartbeat`]. Runs until `shutdown` fires. Gated behind `RUN_BACKGROUND`
    /// in `main` so the bare scaffold serves without driving an election with no peers.
    pub async fn run(&self, mut shutdown: tokio::sync::watch::Receiver<bool>) -> Result<()> {
        let _ = (
            &self.cfg.peers,
            self.cfg.election_timeout,
            self.cfg.heartbeat,
            self.cfg.node_id.as_str(),
        );
        let _ = shutdown.changed().await;
        todo!("V1: run the election/heartbeat loop (randomized timeout, single leader per term)")
    }
}

File: projects/17-global-conferencing/src/global_conferencing/placement.py
"""V1 — Global room placement via consensus. `src/global_conferencing/placement.py`.

This is the control plane every SFU in the mesh must agree on. When a room is
first created, *some* region is chosen as its **home** — the anchor where the
publisher's origin media lives and where the cascade tree (V2/V3) roots — and
that decision must be **identical on every node**, or two people creating
`standup-42` from Tokyo and Frankfurt at the same instant get two disjoint
conferences under one id: a *split room*, conferencing's split-brain. Deciding
one value, cluster-wide, under concurrency and partition, is a **consensus**
problem — exactly what your Raft work in **project 09** is for.

You are **leaning on** project 09's ideas — a randomized election timeout to
elect a leader, append-entries to replicate, a commit index, idempotent apply —
not rebuilding Raft from nothing. What is *yours* here is modelling **room
placement + region membership** as the replicated state machine (the entries in
`rpc.py`) and getting the safety property right: **at most one home region per
room, cluster-wide, forever** — and a minority partition that *cannot* invent one.

## Why this is not a cache, even inside one process

There is no lock in this class and no thread, because one event loop runs every
handler. That does not make it race-free. Every `await` is a point where another
coroutine runs, so

    if room_id not in self._rooms:
        await <replicate a PlaceRoom>        # <- two publishes both get here
        self._rooms[room_id] = ...

lets two concurrent `publish` requests *on the same node* both pass the check and
both propose a home. That is the split room, reproduced without a network. What
serializes the decision — a log index, an `asyncio.Lock`, both — and which of
those you are relying on for *cross-node* safety, belongs in `docs/17-design.md`.

## The two consistency tiers

Placement is strongly consistent; the media riding on it (V2) is best-effort and
drops packets freely. Hot routing reads — `room()`, `snapshot()` — come from the
**locally applied** map, never a round-trip to the leader: a subscribe in Tokyo
must not block on Frankfurt. Those reads are wired below. Everything that
*changes* the map is the V1 worklist.

Scaffold state: construction and the read-only views are wired so `/status` and
`GET /rooms` answer with an empty map. The first `publish` reaches `place_room`
and raises — that is the worklist, rendered by the server as a 501.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, field_serializer

from .cluster import ClusterClient
from .config import PeerNode
from .rpc import ReplicateReply, ReplicateRequest, VoteReply, VoteRequest

__all__ = ["Placement", "PlacementConfig", "Role", "RoomPlacement"]


class Role(StrEnum):
    """This node's role in the placement group. The same three states as project 09."""

    FOLLOWER = "follower"
    CANDIDATE = "candidate"
    LEADER = "leader"

    @property
    def gauge_value(self) -> int:
        """Encoding for the `conf_node_role` gauge: 0 follower, 1 candidate, 2 leader."""
        return {Role.FOLLOWER: 0, Role.CANDIDATE: 1, Role.LEADER: 2}[self]


class RoomPlacement(BaseModel):
    """The committed placement of one room — the state machine's value for it.

    Every node converges to the same `RoomPlacement` for a given `room_id`. Frozen,
    because a committed placement is a fact: applying an entry produces a *new*
    value (`model_copy(update=...)`) rather than editing one another coroutine
    may be holding mid-serialization.
    """

    model_config = ConfigDict(frozen=True)

    room_id: str
    home_region: str
    """The one home region, chosen once by consensus. The invariant V1 protects."""

    active_regions: frozenset[str] = frozenset()
    """Regions with ≥1 live participant — the cascade topology is derived from this."""

    epoch: int = 0
    """Bumped on each committed change to this room (membership included). Re-placing
    an already-placed room must not bump it — that is the "no epoch churn" criterion."""

    @field_serializer("active_regions")
    def _regions_sorted(self, regions: frozenset[str]) -> list[str]:
        # A set has no order; JSON does. Sorting here means two nodes holding the
        # same placement render byte-identical JSON, which is what lets a test
        # (or you, eyeballing three `/rooms` responses) compare them directly.
        return sorted(regions)


@dataclass(frozen=True, slots=True)
class PlacementConfig:
    """The placement group's fixed parameters, built from `Settings` in `main`."""

    region: str
    """This node's region — the home it proposes for rooms created locally."""

    node_id: str
    """This node's stable id, used in vote records and logs."""

    peers: tuple[PeerNode, ...]
    """The other SFUs in the mesh."""

    election_timeout: float
    """Seconds. A follower stands after a random interval in
    `[election_timeout, 2 * election_timeout)` — project 09's split-vote breaker."""

    heartbeat: float
    """Seconds between leader heartbeats; below `election_timeout` (checked in config)."""

    max_rooms: int
    """Cluster-wide cap on placed rooms, enforced through the log — not per node."""

    @property
    def cluster_size(self) -> int:
        return len(self.peers) + 1

    @property
    def quorum(self) -> int:
        """Votes needed to decide anything: a strict majority of the mesh, self included.

        A node that cannot gather this many answers is in a minority and must refuse
        to place a room.
        """
        return self.cluster_size // 2 + 1


class Placement:
    """The placement control plane: a Raft-lite replicated map of room → home + membership."""

    def __init__(self, config: PlacementConfig, cluster: ClusterClient) -> None:
        self.config = config
        self._cluster = cluster
        self._role = Role.FOLLOWER
        self._term = 0
        self._leader: str | None = None
        self._rooms: dict[str, RoomPlacement] = {}
        """The applied state machine: room id → committed placement. Rebuilt by
        applying committed log entries in order; the source of truth for routing reads."""

        # TODO(V1): the rest of the replicated state. Roughly: the log itself
        # (a `list` of `PlacementEntry`, where the index *is* the position), the
        # commit index and the last-applied index, who this node voted for this
        # term, and something the replicate handler can poke to reset the election
        # timer (an `asyncio.Event` is one shape). Project 09 already taught you
        # which of these must survive a restart — the "a node that rejoins learns
        # the committed placements" criterion is where that comes due.

    # ---- wired read views --------------------------------------------------

    @property
    def role(self) -> Role:
        return self._role

    @property
    def term(self) -> int:
        return self._term

    @property
    def leader(self) -> str | None:
        """Who this node believes leads the placement group, if anyone."""
        return self._leader

    def snapshot(self) -> list[RoomPlacement]:
        """Every placed room, from the locally applied map (for `/rooms` + `/status`)."""
        return [self._rooms[room_id] for room_id in sorted(self._rooms)]

    def room(self, room_id: str) -> RoomPlacement | None:
        """The committed placement for one room, if placed — the hot routing read."""
        return self._rooms.get(room_id)

    # ---- V1 worklist: elect · replicate · apply ----------------------------

    async def place_room(self, room_id: str) -> RoomPlacement:
        """Ensure `room_id` has a home region, cluster-wide, and return its placement.

        TODO(V1): if the room is already placed, return that placement — no second
        home, no epoch bump (idempotent). Otherwise propose a `PlaceRoom` entry with
        this node's region as home and drive it to commit.

        * Not the leader? Forward to it, or raise `NotLeaderError(leader)` so
          signaling can — your call, and a line in the design doc.
        * Cannot reach `config.quorum`? Raise `UnavailableError`. Refusing is the
          whole safety property: a minority that "just picks a home for now" is how
          a split room is born, and it is born silently.
        * Enforce `config.max_rooms` **at commit/apply time** (raise
          `ConflictError` to the proposer). A per-node counter checked before
          proposing lets two nodes each admit the last slot.
        * Bump `ROOMS_PLACED` / `PLACEMENT_COMMITS_TOTAL` when the entry applies.
        """
        raise NotImplementedError(
            "V1: place the room via consensus (idempotent, one home, minority refuses)"
        )

    async def register_interest(self, room_id: str, region: str, *, joined: bool) -> None:
        """Replicate that `region` gained its first / lost its last participant in `room_id`.

        TODO(V1): commit a `RegionInterest` entry so every node's `active_regions`
        for the room converges, then apply it (new `RoomPlacement`, epoch + 1).
        Idempotent in the same way placement is: a region joining a room it is
        already active in must not bump the epoch. Leaving drops the region from
        the set — and that change is what lets V2 tear a relay leg down.

        Note the signature takes a *region*, not a participant. Counting local
        participants so that only the first join and the last leave reach the log
        is a local concern; replicating every join would put the join storm in the
        boss fight straight onto the consensus path.
        """
        raise NotImplementedError("V1: replicate the membership change, apply on commit")

    async def on_vote(self, request: VoteRequest) -> VoteReply:
        """Answer a peer's `POST /cluster/vote`.

        TODO(V1): adopt a higher term (and step down to follower if you were
        anything else), grant at most one vote per term, and deny a candidate
        whose log is behind yours — which needs the fields `rpc.py` tells you to
        add. Reply with your current term either way. Update `NODE_TERM` /
        `NODE_ROLE`.
        """
        raise NotImplementedError(
            "V1: RequestVote — grant at most one vote per term, adopt higher terms"
        )

    async def on_replicate(self, request: ReplicateRequest) -> ReplicateReply:
        """Answer the leader's `POST /cluster/replicate` (an empty `entries` is a heartbeat).

        TODO(V1): reject a stale term; otherwise record the leader, reset the
        election timer, append the entries (after the consistency check your
        extended request enables), and apply whatever is newly committed to
        `self._rooms` in log order. Apply must be idempotent — a retried replicate
        delivers the same entries twice, and applying a `PlaceRoom` twice must not
        change a committed home.
        """
        raise NotImplementedError(
            "V1: AppendEntries — append/commit entries, apply to the map, reset election timer"
        )

    async def run(self) -> None:
        """The election + heartbeat loop. Runs until cancelled.

        TODO(V1): as a follower, wait a random interval in
        `[election_timeout, 2 * election_timeout)` (`random.uniform`) for a
        heartbeat; if none arrives, become candidate, bump the term, and canvass
        `self._cluster.regions` concurrently (see `cluster.py` on why concurrently,
        and on why `PeerUnreachableError` is an ordinary non-vote). As leader, send
        `replicate` to every peer every `config.heartbeat`. Bump `ELECTIONS_TOTAL`
        per election started.

        Cancellation is the shutdown signal — `main` cancels this task on SIGTERM.
        Do not swallow `asyncio.CancelledError` in a broad `except`: catch
        `PeerUnreachableError`, not `Exception`, or the loop will outlive shutdown.

        Gated behind `RUN_BACKGROUND` so the bare scaffold boots without driving an
        election with no peers.
        """
        raise NotImplementedError(
            "V1: run the election/heartbeat loop (randomized timeout, single leader per term)"
        )

    async def step_down(self) -> None:
        """Relinquish leadership on graceful shutdown, so the mesh re-elects fast.

        TODO(V1): a no-op unless this node leads. A leader that simply vanishes
        costs the mesh a full election timeout with no leader — and every
        `publish` for a new room in that window refuses. How you shorten that
        window (and what a follower should make of it) is a design-doc line.
        Called from the lifespan's shutdown, before the cluster client closes.
        """
        raise NotImplementedError("V1: relinquish leadership if held (graceful shutdown)")
